idr/marshal2.go

```diff
@@ -0,0 +1,205 @@
+	// contains an array of one element which is 'child'. There is really no other signs/flags
+	// help us to disambiguate the situation. So use common sense here, treat this situation
+	// like a single child field, instead of a child array with one element. EDI is similar.
+	// EDI doesn't have some indicator (i.e. a segment can have multiple instances), but it's
+	// too hard to pass the indicator here. So hope this promise/limitation is acceptable.
+	elemNum := 0
+	elemName := (*string)(nil)
+	for c := n.FirstChild; c != nil; c = c.NextSibling {
+		if c.Type != ElementNode {
+			continue
+		}
+		elemNum++
+		if elemName == nil {
+			elemName = strs.StrPtr(j2NodeName(c))
+		} else if j2NodeName(c) != *elemName {
+			// We have differently named child element nodes, clearly
+			// this isn't an array case.
+			return false
+		}
+	}
+	// If the number of identically named child element nodes is > 1 (including JSON's "" named child
+	// element nodes) then it's an array case;
+	// If there is only one child element node, then it is array case only when its name is "" - that's
+	// in JSON array case. Otherwise, just treat it as a single child field case.
+	return elemNum > 1 || (elemNum == 1 && *elemName == "")
+}
+
+func getChildText(n *Node) interface{} {
+	if !IsJSON(n) {
+		return n.InnerText()
+	}
+	n = n.FirstChild
+	switch {
+	case IsJSONValueNum(n):
+		f, _ := strconv.ParseFloat(n.Data, 64)
+		return f
+	case IsJSONValueBool(n):
+		b, _ := strconv.ParseBool(n.Data)
+		return b
+	case IsJSONValueNull(n):
+		return nil
+	default:
+		return n.Data
+	}
+}
+
+// J2NodeToInterface converts *Node into an interface{} that, once marshaled, is JSON friendly.
+func J2NodeToInterface(n *Node) interface{} {
+	switch {
+	case isChildText(n):
+		return getChildText(n)
+	case isChildArray(n):
+		arr := make([]interface{}, 0)
+		for c := n.FirstChild; c != nil; c = c.NextSibling {
+			// unfortunately there doesn't seem to be a way to include AttributeNode elegantly
+			// in an XML array situation:
+			//  <abc attr1="1" attr2="2">
+			//    <e>one</e>
+			//    <e>two</e>
+			//  </abc>
+			// We want to return, for the node of <abc>, this:
+			//  [ "one", "two" ]
+			// It's hard to jam in the attributes collection somewhere without causing confusion.
+			// Don't think it's a prevailing or even relevant case. So be it.
+			if c.Type != ElementNode {
+				continue
+			}
+			arr = append(arr, J2NodeToInterface(c))
+		}
+		return arr
+	default:
+		fields := make(map[string]interface{})
+		attrs := make(map[string]interface{})
+		fieldIsArr := make(map[string]bool)
+		for c := n.FirstChild; c != nil; c = c.NextSibling {
+			name := j2NodeName(c)
+			value := J2NodeToInterface(c)
+			if c.Type == ElementNode {
+				// Both in XML and EDI, we can potentially have something like the following:
+				//   <abc>
+				//     <efg>1</abc>
+				//     <efg>2</abc>
+				//     <xyz>3</efg>
+				//   </abc>
```

Comment:
is this is a correct xml, should we need to deal with this?

Reply:
fixed